File: src/lib/email.ts
import nodemailer from 'nodemailer';

// Create a transporter object
let transporter: nodemailer.Transporter | null = null;

// Initialize the transporter
export async function initTransporter() {
  // For development, use a test account
  console.warn('Email credentials not found, using test account');
  try {
    const testAccount = await nodemailer.createTestAccount();
    transporter = nodemailer.createTransport({
      host: 'smtp.hostinger.com',
      port: 587,
      secure: false,
      auth: {
        user: 'naveed@codehuntspk.com',
        pass: '',
      },
    });
    console.log('Test account created:', testAccount.user);
    console.log('Test account password:', testAccount.pass);
  } catch (error) {
    console.error('Failed to create test account:', error);
    throw error;
  }
}

// Send an email
export async function sendEmail({
                                  to,
                                  subject,
                                  html,
                                  text,
                                }: {
  to: string;
  subject: string;
  html: string;
  text: string;
}) {
  // Ensure transporter is initialized
  if (!transporter) {
    console.log('Transporter not initialized, initializing now...');
    await initTransporter();
  }

  if (!transporter) {
    throw new Error('Failed to initialize email transporter');
  }

  try {
    const info = await transporter.sendMail({
      from: '"CryptoForex" <naveed@codehuntspk.com>',
      to,
      subject,
      text,
      html,
    });

    console.log('Message sent: %s', info.messageId);

    // For development, log the test URL
    console.log('Preview URL: %s', nodemailer.getTestMessageUrl(info));

    return { success: true, messageId: info.messageId };
  } catch (error) {
    console.error('Error sending email:', error);
    return { success: false, error };
  }
}

// Send a verification email
export async function sendVerificationEmail(
    email: string,
    token: string,
    baseUrl: string
) {
  const verificationUrl = `${baseUrl}/verify-email?token=${token}`;

  return sendEmail({
    to: email,
    subject: 'Verify your email address',
    text: `Please verify your email address by clicking on the following link: ${verificationUrl}`,
    html: `
      <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
        <h2 style="color: #333;">Verify your email address</h2>
        <p>Thank you for signing up with CryptoForex. Please verify your email address by clicking on the button below:</p>
        <div style="text-align: center; margin: 30px 0;">
          <a href="${verificationUrl}" style="background-color: #4F46E5; color: white; padding: 12px 24px; text-decoration: none; border-radius: 4px; display: inline-block;">Verify Email</a>
        </div>
        <p>If the button doesn't work, you can also copy and paste the following link into your browser:</p>
        <p style="word-break: break-all; color: #4F46E5;">${verificationUrl}</p>
        <p>This link will expire in 24 hours.</p>
        <p>If you didn't sign up for CryptoForex, you can safely ignore this email.</p>
      </div>
    `,
  });
}

// Send a password reset email
export async function sendPasswordResetEmail(
    email: string,
    token: string,
    baseUrl: string
) {
  console.log("I M GOING IN")
  const resetUrl = `${baseUrl}/reset-password?token=${token}`;

  return sendEmail({
    to: email,
    subject: 'Reset your password',
    text: `Reset your password by clicking on the following link: ${resetUrl}`,
    html: `
      <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
        <h2 style="color: #333;">Reset your password</h2>
        <p>You requested to reset your password. Click the button below to set a new password:</p>
        <div style="text-align: center; margin: 30px 0;">
          <a href="${resetUrl}" style="background-color: #4F46E5; color: white; padding: 12px 24px; text-decoration: none; border-radius: 4px; display: inline-block;">Reset Password</a>
        </div>
        <p>If the button doesn't work, you can also copy and paste the following link into your browser:</p>
        <p style="word-break: break-all; color: #4F46E5;">${resetUrl}</p>
        <p>This link will expire in 1 hour.</p>
        <p>If you didn't request a password reset, you can safely ignore this email.</p>
      </div>
    `,
  });
}

File: backend/app/services/email.py
"""
Email service for sending notifications and verification emails
"""

import smtplib
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from typing import Optional
import logging
from app.core.config import settings

logger = logging.getLogger(__name__)

async def send_email(
    to: str,
    subject: str,
    body: str,
    html_body: Optional[str] = None
) -> bool:
    """
    Send an email to a recipient
    
    Args:
        to: Recipient email address
        subject: Email subject
        body: Plain text body
        html_body: Optional HTML body
    
    Returns:
        True if email was sent successfully, False otherwise
    """
    try:
        # Create message
        msg = MIMEMultipart('alternative')
        msg['Subject'] = subject
        msg['From'] = f"{settings.APP_NAME} <{settings.FROM_EMAIL}>"
        msg['To'] = to
        
        # Add plain text part
        part1 = MIMEText(body, 'plain')
        msg.attach(part1)
        
        # Add HTML part if provided
        if html_body:
            part2 = MIMEText(html_body, 'html')
            msg.attach(part2)
        
        # Send email
        if settings.SMTP_HOST and settings.SMTP_USERNAME:
            with smtplib.SMTP(settings.SMTP_HOST, settings.SMTP_PORT) as server:
                if settings.SMTP_PORT == 587:
                    server.starttls()
                if settings.SMTP_PASSWORD:
                    server.login(settings.SMTP_USERNAME, settings.SMTP_PASSWORD)
                server.send_message(msg)
            
            logger.info(f"Email sent successfully to {to}")
            return True
        else:
            logger.warning(f"Email service not configured. Would have sent: {subject} to {to}")
            return True  # Return True in development
            
    except Exception as e:
        logger.error(f"Failed to send email to {to}: {str(e)}")
        return False

async def send_verification_email(email: str, token: str) -> bool:
    """
    Send a verification email to a new user
    
    Args:
        email: User's email address
        token: Verification token
    
    Returns:
        True if email was sent successfully, False otherwise
    """
    subject = f"Welcome to {settings.APP_NAME} - Verify Your Email"
    
    verification_link = f"http://localhost:3000/verify-email?token={token}"
    
    body = f"""
Welcome to {settings.APP_NAME}!

Please verify your email address by clicking the link below:

{verification_link}

This link will expire in 24 hours.

If you didn't create an account with us, please ignore this email.

Best regards,
The {settings.APP_NAME} Team
"""
    
    html_body = f"""
<!DOCTYPE html>
<html>
<head>
    <style>
        .container {{
            font-family: Arial, sans-serif;
            max-width: 600px;
            margin: 0 auto;
            padding: 20px;
        }}
        .header {{
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            color: white;
            padding: 30px;
            text-align: center;
            border-radius: 10px 10px 0 0;
        }}
        .content {{
            background: #f7f7f7;
            padding: 30px;
            border-radius: 0 0 10px 10px;
        }}
        .button {{
            display: inline-block;
            padding: 12px 30px;
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            color: white;
            text-decoration: none;
            border-radius: 5px;
            margin: 20px 0;
        }}
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>Welcome to {settings.APP_NAME}!</h1>
        </div>
        <div class="content">
            <p>Thank you for joining our platform.</p>
            <p>Please verify your email address by clicking the button below:</p>
            <center>
                <a href="{verification_link}" class="button">Verify Email</a>
            </center>
            <p>Or copy and paste this link into your browser:</p>
            <p style="word-break: break-all; color: #666;">{verification_link}</p>
            <p>This link will expire in 24 hours.</p>
            <p>If you didn't create an account with us, please ignore this email.</p>
            <hr style="border: none; border-top: 1px solid #ddd; margin: 30px 0;">
            <p style="color: #666; font-size: 12px;">
                Best regards,<br>
                The {settings.APP_NAME} Team
            </p>
        </div>
    </div>
</body>
</html>
"""
    
    return await send_email(email, subject, body, html_body)

async def send_password_reset_email(email: str, token: str) -> bool:
    """
    Send a password reset email
    
    Args:
        email: User's email address
        token: Reset token
    
    Returns:
        True if email was sent successfully, False otherwise
    """
    subject = f"{settings.APP_NAME} - Password Reset Request"
    
    reset_link = f"http://localhost:3000/reset-password?token={token}"
    
    body = f"""
You have requested to reset your password.

Please click the link below to reset your password:

{reset_link}

This link will expire in 1 hour.

If you didn't request this, please ignore this email and your password will remain unchanged.

Best regards,
The {settings.APP_NAME} Team
"""
    
    return await send_email(email, subject, body)

async def send_deposit_confirmation_email(email: str, amount: float, currency: str, tx_id: str) -> bool:
    """
    Send a deposit confirmation email
    
    Args:
        email: User's email address
        amount: Deposit amount
        currency: Currency code
        tx_id: Transaction ID
    
    Returns:
        True if email was sent successfully, False otherwise
    """
    subject = f"{settings.APP_NAME} - Deposit Confirmation"
    
    body = f"""
Your deposit has been confirmed!

Details:
- Amount: {amount} {currency}
- Transaction ID: {tx_id}
- Status: Completed

The funds are now available in your account.

Best regards,
The {settings.APP_NAME} Team
"""
    
    return await send_email(email, subject, body)

async def send_withdrawal_confirmation_email(email: str, amount: float, currency: str, tx_id: str) -> bool:
    """
    Send a withdrawal confirmation email
    
    Args:
        email: User's email address
        amount: Withdrawal amount
        currency: Currency code
        tx_id: Transaction ID
    
    Returns:
        True if email was sent successfully, False otherwise
    """
    subject = f"{settings.APP_NAME} - Withdrawal Processed"
    
    body = f"""
Your withdrawal has been processed!

Details:
- Amount: {amount} {currency}
- Transaction ID: {tx_id}
- Status: Completed

The funds have been sent to your designated address/account.

Best regards,
The {settings.APP_NAME} Team
"""
    
    return await send_email(email, subject, body)
